### PR TITLE
Allow Users to Create Tournaments

### PR DESCRIPTION
**Pull request by freefal**
Closes https://github.com/veloce/lichobile/issues/364

### CODE REVIEW

File: src/js/ui/tournament/tournamentXhr.js [hunk 1]
```diff
@@ -27,3 +27,18 @@ export function withdraw(id) {
 export function playerInfo(tournamentId, playerId) {
   return fetchJSON('/tournament/' + tournamentId + '/player/' + playerId, {}, true);
 }
+
+export function create(variant, position, mode, clockTime, clockIncrement, minutes, waitMinutes) {
```
Comment (veloce) [inline src/js/ui/tournament/tournamentXhr.js]: `fetchJSON` returns a `Promise<T>` it's good to type the xhr function

File: src/js/ui/shared/form.ts [hunk 2]
```diff
@@ -14,15 +14,17 @@ function renderOption(label: string, value: string, storedValue: string, labelAr
 }
 
 
-function renderOptionGroup(label, value, storedValue, labelArg, labelArg2) {
+function renderOptionGroup(label:string, value:any, storedValue:string, labelArg:string, labelArg2:string): any {
```
Comment (veloce) [inline src/js/ui/shared/form.ts]: Why `any` for value? it should be `string`, or maybe `string | number`, but not `any`

File: src/js/ui/shared/form.ts [hunk 3]
```diff
@@ -111,14 +113,20 @@ export default {
...
-    const onChange = function(e) {
-      settingsProp(e.target.value);
-      if (onChangeCallback) onChangeCallback(e.target.value);
-      setTimeout(function() {
-        redraw();
-      }, 10);
+  renderSelectWithGroup(
+    label: string,
+    name: string,
+    options: any,
```
Comment (veloce) [inline src/js/ui/shared/form.ts]: no `any` here; in the other select I use `Array<SelectOption>`

**Comment by freefal** [comment]
Yes. Apologies. I marked this PR "[WIP]" because I know there are some
hacks in it. Just wanted to get back to something that compiled first.

**Comment by veloce** [comment]
Ah, ok sure. I'll wait for the [WIP] removal

File: src/js/ui/shared/form.ts [hunk 4]
```diff
@@ -13,6 +14,21 @@ function renderOption(label: string, value: string, storedValue: string, labelAr
   }, i18n(label, labelArg, labelArg2));
 }
 
+
+function renderOptionGroup(label:string, value:string | SelectOptionGroup, storedValue:string, labelArg:string, labelArg2:string): Mithril.Children {
+  if (typeof value === 'string') {
```
Comment (veloce) [inline src/js/ui/shared/form.ts]: You should use a **typeGuard** instead, this is a typescript feature explained here: http://www.typescriptlang.org/docs/handbook/advanced-types.html

File: src/js/ui/tournament/detail/tournamentDetailCtrl.ts [hunk 5]
```diff
@@ -8,20 +8,21 @@ import * as helper from '../../helper';
 import * as m from 'mithril';
 import faq from '../faq';
 import playerInfo from '../playerInfo';
+import { TournamentAttrs, Tournament, FeaturedGameUpdate, TournamentState } from '../interfaces'
 
-export default function oninit(vnode) {
+export default function oninit(vnode: Mithril.Vnode<TournamentAttrs>) {
```
Comment (veloce) [inline src/js/ui/tournament/detail/tournamentDetailCtrl.ts]: Correct type is `Mithril.Vnode<Attrs, State>`

File: src/js/ui/tournament/detail/tournamentDetailView.tsx [hunk 6]
```diff
@@ -10,41 +10,42 @@ import playerInfo from '../playerInfo';
 import * as helper from '../../helper';
 import settings from '../../../settings';
 import miniBoard from '../../shared/miniBoard';
+import { TournamentState, Tournament, PlayerInfoState, StandingPlayer, PodiumPlace } from '../interfaces';
 
-export default function view(vnode) {
-  const ctrl = vnode.state;
+export default function view(vnode: Mithril.Vnode<{}>) {
```
Comment (veloce) [inline src/js/ui/tournament/detail/tournamentDetailView.tsx]: fix type

File: src/js/ui/tournament/tournamentCtrl.ts [hunk 7]
```diff
@@ -5,14 +5,15 @@ import * as xhr from './tournamentXhr';
 import * as helper from '../helper';
 import * as m from 'mithril';
 import settings from '../../settings';
+import { TournamentListAttrs, TournamentListItem, TournamentListsState, TournamentLists } from './interfaces';
 
-export default function oninit(vnode) {
+export default function oninit(vnode: Mithril.Vnode<TournamentListAttrs>) {
```
Comment (veloce) [inline src/js/ui/tournament/tournamentCtrl.ts]: fix type

File: src/js/ui/tournament/tournamentView.tsx [hunk 8]
```diff
@@ -6,12 +6,15 @@ import layout from '../layout';
 import i18n from '../../i18n';
 import * as m from 'mithril';
 import tabs from '../shared/tabs';
+import newTournamentForm from './newTournamentForm';
+import { TournamentListsState, TournamentListItem } from './interfaces';
 
-export default function view(vnode) {
-  const ctrl = vnode.state;
+export default function view(vnode: Mithril.Vnode<{}>) {
```
Comment (veloce) [inline src/js/ui/tournament/tournamentView.tsx]: fix type

File: src/js/ui/shared/form.ts (hunk 4 above)
Comment (freefal) [inline src/js/ui/shared/form.ts]: Looks like typeGuard can be implemented with a function or simply allowing typescript to infer types from the `typeof` if statements. I removed the manual casting and the latter approach seems to work great. Good to know.

File: src/js/ui/tournament/tournamentView.tsx (hunk 8 above)
Comment (freefal) [inline src/js/ui/tournament/tournamentView.tsx]: I fixed all Mithril types as well.

File: src/js/ui/shared/form.ts (hunk 4 above)
Comment (veloce) [inline src/js/ui/shared/form.ts]: that's right, type guards are simple predicate functions with a specific return type annotation

**Pull request merged**